### PR TITLE
Warn admin on login if their password is still "admin"

### PR DESCRIPTION
**Pull request by xmo-odoo**
Task 31122 Part 1 (though the last one), replaces #22428

Go through private channels (suggestion of @KangOl) as we want an immediate but not too obtrusive notification which nonetheless survives a redirection. message_post might go through an email which is undesirable.

### CODE REVIEW

File: addons/web/controllers/main.py [hunk 1]
```diff
@@ -479,6 +493,10 @@ def web_login(self, redirect=None, **kw):
             old_uid = request.uid
             try:
                 uid = request.session.authenticate(request.session.db, request.params['login'], request.params['password'])
+                if request.params['password'] == 'admin':
+                    admin = request.env.ref('base.partner_admin')
+                    if uid in admin.user_ids.ids:
+                        _password_warn(request.env(user=1), admin)
```
Comment (KangOl) [inline addons/web/controllers/main.py]: use `SUPERUSER_ID` constant.

File: addons/web/controllers/main.py [hunk 2]
```diff
@@ -424,6 +424,20 @@ def xml2json_from_elementtree(el, preserve_whitespaces=False):
     res["children"] = kids
     return res
 
+def _password_warn(env, partner):
+    """ Admin still has `admin` password, flash a message via chatter
+
+    Does not use message_post because depending on conf' that might send an
```
Comment (odony) [inline addons/web/controllers/main.py]: Is it me or you're actually using `message_post` now? ;-)

File: addons/web/controllers/main.py [hunk 3]
```diff
@@ -424,6 +424,20 @@ def xml2json_from_elementtree(el, preserve_whitespaces=False):
...
+    email instead.
+    """
+    MailChannel = env['mail.channel']
+    MailChannel.browse(MailChannel.channel_get([partner.id])['id'])\
+        .message_post(
+            body=_("Your administrator password is the default (admin), if this is any sort of production database you may want to change that for security reasons."),
```
Comment (odony) [inline addons/web/controllers/main.py]:
```suggestion
            body=_("Your password is the default (admin)! If this system is exposed to untrusted users it is important to change it immediately for security reasons. I will keep nagging you about it!"),
```

Rationale: it is important to do it _also_ for test/dev databases when they are exposed publicly. They can be abused too in various ways.

**Comment by odony** [comment]
Other notes, for the record:
- We could deactivate the warning in `--dev` mode too, but I don't think that's a good idea. Many devs don't use it, while some admins *do* use it on public-facing systems. The heuristic with private IPs seems wiser.
- It's gonna be annoying on runbot, but I don't think we can avoid that easily. Having an ICP kill flag would require non-standard DB provisioning, a bad precedent IMO.

File: addons/web/controllers/main.py (hunk 2 above)
Comment (xmo-odoo) [reply addons/web/controllers/main.py]: Kinda yeah, that was about using mail thread's message_post directly.

**Comment by xmo-odoo** [comment]
> Shouldn't _password_warn() be a private method on res.users, so it can be inherited, etc? Let's say someone hypothetically wants to customize this behavior... ;-)

I don't know that there's a use case for it, and MFA will eventually require splitting the "interactive" (web) and "non-interactive" (API) auths completely — though I'm not quite sure how yet — so that'll provide the relevant customisation hook.

> I think we should warn any user that is a member of base.group_system / base.group_erp_manager, not just the default admin. Even if they're not the original admin, using "admin" as the password is ill-advised. Might be tempting to do it for any user, but that verges on a generic password-strength enforcement.

"admin" is the default password for the admin. Other users have to be created explicitly and already have a password strength meter thing *and* a requirement of 8 characters by default. More checking of non-special / non-default cases seems unnecessary.

> In order to avoid a constant annoyance for devs, shouldn't we bypass the warning when the request is coming from an RFC 1918 private IP?

Easy enough given [the ipaddress module](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_private).

**Comment by xmo-odoo** [comment]
robodoo r+

**Comment by mart-e** [comment]
Looks like `is_private` is not properly detected. On runbot we should be spammed with this message but acutally we aren't. Just noticed this warning message on [this build](https://1953028-master-3b40c6.runbot39.odoo.com/) (but not sure why on this one on particular and not all the other runbot...)
My guess is that `is_private` will be `True` if the ip is not properly forwarded and will depends of the nginx config/werkzeug.

How about using the `demo` value of the db to decide?

**Pull request closed without merging**